builder.go

```diff
@@ -0,0 +1,248 @@
+package pitaya
+
+import (
+	"github.com/google/uuid"
+	"github.com/spf13/viper"
+	"github.com/topfreegames/pitaya/acceptor"
+	"github.com/topfreegames/pitaya/agent"
+	"github.com/topfreegames/pitaya/cluster"
+	"github.com/topfreegames/pitaya/config"
+	"github.com/topfreegames/pitaya/conn/codec"
+	"github.com/topfreegames/pitaya/conn/message"
+	"github.com/topfreegames/pitaya/defaultpipelines"
+	"github.com/topfreegames/pitaya/groups"
+	"github.com/topfreegames/pitaya/logger"
+	"github.com/topfreegames/pitaya/metrics"
+	"github.com/topfreegames/pitaya/pipeline"
+	"github.com/topfreegames/pitaya/router"
+	"github.com/topfreegames/pitaya/serialize"
+	"github.com/topfreegames/pitaya/serialize/json"
+	"github.com/topfreegames/pitaya/service"
+	"github.com/topfreegames/pitaya/session"
+	"github.com/topfreegames/pitaya/worker"
+)
+
+// Builder holds dependency instances for a pitaya App
+type Builder struct {
+	acceptors        []acceptor.Acceptor
+	Configs          []*viper.Viper
+	DieChan          chan bool
+	PacketDecoder    codec.PacketDecoder
+	PacketEncoder    codec.PacketEncoder
+	MessageEncoder   *message.MessagesEncoder
+	Serializer       serialize.Serializer
+	Router           *router.Router
+	RPCClient        cluster.RPCClient
+	RPCServer        cluster.RPCServer
+	MetricsReporters []metrics.Reporter
+	Server           *cluster.Server
+	ServerMode       ServerMode
+	ServiceDiscovery cluster.ServiceDiscovery
+	Groups           groups.GroupService
+	SessionPool      session.SessionPool
+	Worker           *worker.Worker
+	HandlerHooks     *pipeline.HandlerHooks
+}
+
+// PitayaBuilder Builder interface
+type PitayaBuilder interface {
```

Comment:
Does the builder needs to be an interface as well?

Reply:
Good point, it doesn't. I probably forgot to remove it.
Also, btw, I just created this builder to keep functionality as it was before, easily creating pitaya with lots of default stuff, but also giving flexibility to customize implementations.

In general, I don't like this "default" approach, and in my use case I ended up not using it. So we might want to consider removing the builder altogether and using the explicit NewApp constructor with all required dependencies.

The downside is having to explicitly instantiate all dependencies.

Reply:
I think having a builder is still desirable because it allows you to quickly create a server for testing/prototyping. Since it is an open-source project, we should have an easy way to create a server.